litex/soc/integration/doc.py

```diff
@@ -0,0 +1,150 @@
+# This file is Copyright (c) 2019 Sean Cross <sean@xobs.io>
+
+from migen.fhdl.module import DUID
+from migen.util.misc import xdir
+
+from litex.soc.interconnect.csr_eventmanager import EventManager
+
+import textwrap
+import inspect
+
+class ModuleDoc(DUID):
+    """Module Documentation Support
+
+    ModuleDoc enables you to add documentation to your Module.  This documentation is in addition to
+    any CSR-level documentation you may add to your module, if applicable.
+
+    There are two ways to use :obj:`ModuleDoc`:
+
+    1. Inherit :obj:`ModuleDoc` as part of your class.  The docstring of your class will become the
+    first section of your class' module documentation
+    2. Add a :obj:`ModuleDoc` object to your class and inherit from :obj:`AutoDoc`.
+
+    If you inherit from :obj:`ModuleDoc`, then there is no need to call ``__init__()``
+
+    Synopsis
+    --------
+
+    ::
+
+        class SomeClass(Module, ModuleDoc, AutoDoc):
+            \"\"\"Some Special Hardware Module
+
+            This is a hardware module that implements something really cool.
+            \"\"\"
+
+            def __init__(self):
+                self.other_section = ModuleDoc(title="Protocol Details", body="This section details more
+                information about the protocol")
+
+    """
+
+    def __init__(self, body=None, title=None, file=None, format="rst"):
+        """Construct a :obj:`ModuleDoc` object for use with :obj:`AutoDoc`
+
+        Arguments
+        ---------
+
+        body (:obj:`str`): Main body of the document.  If ``title`` is omitted, then the
+        title is taken as the first line of ``body``.
+
+        title (:obj:`str` Optional): Title of this particular section.
+
+        file (:obj:`str` Optional): It is possible to load the documentation from an external
+        file instead of specifying it inline.  This allows for the use of an external text
+        editor.  If a ``file`` is specified, then it will override the ``body`` argument.
+
+        format (:obj:`str` Optional): The text format.  Python prefers reStructured Text, so this
+        defaults to `rst`.  If specifying a `file`, then the suffix will be used instead of
+        `format`.  If you specify a format other than `rst`, you may need to install a converter.
+        """
+
+        import os

```

Comment:
Shouldn't import inside a function.

Reply:
Why not?  Is there a `Python Best Practices` I could read that has information about things like this?

Reply:
pylint / pep8 will normally complain about this.

https://www.python.org/dev/peps/pep-0008/#imports